core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -211,32 +230,49 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
         rawConnectivityState = IDLE;
         updateBalancingState(IDLE, new RequestConnectionPicker(this));
         break;
+
       case CONNECTING:
         rawConnectivityState = CONNECTING;
         updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
         break;
+
       case READY:
         shutdownRemaining(subchannelData);
         addressIndex.seekTo(getAddress(subchannel));
         rawConnectivityState = READY;
         updateHealthCheckedState(subchannelData);
         break;
+
       case TRANSIENT_FAILURE:
         // If we are looking at current channel, request a connection if possible
         if (addressIndex.isValid()
             && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
-          addressIndex.increment();
-          requestConnection();
+          if (addressIndex.increment()) {
```

Comment:
There are 2 other cases where we hit the end  of addressIndex and for both of them we don't know that a connection attempt has completed:

- the happy eyeballs scheduled task's deadline being exceeded (`StartNextConnection.run`)
- `RequestConnectionPicker.pickSubchannel`

Reply:
1. You are right. I see that the "first pass complete" check I proposed is not sufficient in the case where they are incremented somewhere else.
One concern I have is that in those cases where the first pass is done happened in other places instead of `processSubchannelState`, we are checking `firstPass` at the next subchannel failure, which is too late, worse case is that the transport comes back until it times out.  I think the way C++ is doing is more accurate, https://github.com/grpc/grpc/blob/6a4b5ccea362dd47562c698dd2aa5a8aead41a6a/src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc#L1150 what do you think?
2. This is minor but you could have let the numTF only increment when first pass is done. But either way.
3. Add `enableHappyEyeballs` condition.

Reply:
1. We need to check when the connection attempt completes which is when it calls processSubchannelState, so I believe that is the appropriate way to do it.  I've updated the logic to actually look at the subchannel states to determine that they have all completed.
2. I believe that the new logic using passComplete() does this
3. The old PickFirstLoadBalancer calls refreshNameResolution on every TF it receives [code](https://github.com/grpc/grpc-java/blob/be74bbdfe0d8a799e7fc32155bb0772371ab8b4c/core/src/main/java/io/grpc/internal/PickFirstLoadBalancer.java#L116-L118).  I agree with an earlier comment you made that name resolution should be refreshed like in the happy eyeballs case only at end of first pass and after address count worth of TF status reports, though @ejona86 may disagree.

Reply:
1.   When `requestConnection` from `pick()` or immediate after `acceptResolvedAddress` reaches the last address, it won't call back `processSubchannelState`,  e.g. when all subchannel in TF status already.  I believe the behaviour won't be a huge disaster but still different.
3. The old PickFirstLoadBalancer only has one subchannel but the iteration logic is inside one subchannel, so it calls refreshNameResolution every TF actually means every iteration of all addresses. It looks before your happy eyeballs change it only refreshNameResolution once, which seems a bug. I'm ok with your change this way unless Eric has different views.

Reply:
I think this still remains an issue though:

> When requestConnection from pick() or immediately after acceptResolvedAddress reaches the last address in its iteration, `processSubchannelState` does not come back, e.g. when all subchannel in TF status already. This would cause the refreshNameResolution delayed.

If you have different opinion, let's engage Eric in the thread. Otherwise at least add comment to fixit.

Reply:
I created a couple of tests `happy_eyeballs_connection_results_happen_after_get_to_end` and `happy_eyeballs_pick_pushes_index_over_end` which try these use cases and they don't have any issues.